File: 프로그래머스/2/389480. 완전범죄/완전범죄.py
def solution(info, n, m):
    answer = 0
    INF=int(1e9)
    dp=[INF]*m
    dp[0]=0
    for a_trace, b_trace in info:
        new_dp=[INF]*m
        for prev_b in range(m):
            if(dp[prev_b]==INF):
                continue
            new_a=dp[prev_b]+a_trace
            if(new_a<n):
                new_dp[prev_b]=min(new_a, new_dp[prev_b])
            new_b=prev_b+b_trace
            if(new_b<m):
                new_dp[new_b]=min(dp[prev_b],new_dp[new_b])
        dp=new_dp
    answer=min(dp)
    if(answer==INF):
        return -1
    return answer

File: 프로그래머스/2/389480. 완전범죄/완전범죄.java
import java.util.*;

class Solution {
    static final int INF = 1000000000;
    public int solution(int[][] info, int n, int m) {
        
        int answer = INF;
        int[] dp=new int[m];
        for(int i=0;i<m;i++){
            dp[i]=INF;
        }
        dp[0]=0;
        
        for(int[] trace: info){
            int a_trace=trace[0];
            int b_trace=trace[1];
            int[] new_dp=new int[m];
            for(int i=0;i<m;i++){
                new_dp[i]=INF;
            }
            for(int prev_b=0; prev_b<m; prev_b++){
                if(dp[prev_b]==INF){
                    continue;
                }
                //A가 훔침
                int a_new=dp[prev_b]+a_trace;
                if(a_new<n){
                    new_dp[prev_b]=Math.min(a_new,new_dp[prev_b]);
                }
                //B가 훔침
                int b_new=prev_b+b_trace;
                if(b_new<m){
                    new_dp[b_new]=Math.min(dp[prev_b],new_dp[b_new]);
                }
            }
            dp=new_dp;
        }
        for(int i=0;i<m;i++){
            if(answer>dp[i]){
                answer=dp[i];
            }
        }
        if(answer==INF){
            return -1;
        }
        return answer;
    }
}